File: Foundation/Site.m
//
//  Site.m
//  PupTent
//
//  (c) 2014 @toddheasley
//

#import "Site.h"

static NSString *kNameKey = @"name";
static NSString *kURIKey = @"URI";
static NSString *kTwitterNameKey = @"twitterName";
static NSString *kDomainKey = @"domain";
static NSString *kPagesKey = @"pages";

@implementation Site

- (NSArray *)indexedPages {
    NSMutableArray *pages = [NSMutableArray array];
    for (Page *page in self.pages) {
        if (page.index) {
            [pages addObject:page];
        }
    }
    return [NSArray arrayWithArray:pages];
}

- (NSArray *)featuredPages {
    NSMutableArray *pages = [NSMutableArray array];
    for (Page *page in self.pages) {
        if (page.feature) {
            [pages addObject:page];
        }
    }
    return [NSArray arrayWithArray:pages];
}

- (id)init {
    self = [super init];
    if (self) {
        self.name = @"";
        self.domain = @"";
        self.URI = @"index.html";
        self.twitterName = @"";
        self.pages = @[
            [[Page alloc] init]
        ];
    }
    return self;
}

- (instancetype)initWithDictionary:(NSDictionary *)dictionary {
    self = [super init];
    if (self) {
        NSMutableArray *pages = [NSMutableArray new];
        for (NSDictionary *page in (NSArray *)[dictionary objectForKey:kPagesKey]) {
            [pages addObject:[[Page alloc] initWithDictionary:page]];
        }
        
        self.name = [dictionary objectForKey:kNameKey];
        self.domain = [dictionary objectForKey:kDomainKey];
        self.URI = [dictionary objectForKey:kURIKey];
        self.twitterName = [dictionary objectForKey:kTwitterNameKey];
        self.pages = [NSArray arrayWithArray:pages];
    }
    return self;
}

- (NSDictionary *)dictionary {
    NSMutableArray *pages = [NSMutableArray new];
    for (Page *page in self.pages) {
        [pages addObject:page.dictionary];
    }
    
    return @{
        kNameKey: self.name,
        kDomainKey: self.domain,
        kURIKey: self.URI,
        kTwitterNameKey: self.twitterName,
        kPagesKey: [NSArray arrayWithArray:pages]
    };
}

- (NSArray *)manifest {
    NSMutableArray *manifest = [NSMutableArray new];
    [manifest addObject:self.URI];
    for (Page *page in self.pages) {
        [manifest addObjectsFromArray:page.manifest];
    }
    return [NSArray arrayWithArray:manifest];
}

- (NSString *)description {
    if (self.domain.length > 0) {
        return [NSString stringWithFormat:@"%@ %@", self.domain, [super description]];
    }
    return [super description];
}

@end

File: Foundation/Site.swift
//
//  Site.swift
//  PupTent
//
//  (c) 2014 @toddheasley
//

import Foundation

public class Site: Archiving {
    public var name: String
    public var URI: String
    public var twitterName: String
    public var domain: String
    public var pages: Array<Page>
    public var indexedPages: Array<Page> {
        get {
            var pages: Array<Page> = []
            for page in self.pages {
                if (page.index) {
                    pages.append(page)
                }
            }
            return pages
        }
    }
    
    public convenience init() {
        self.init(dictionary: [
            ArchivingKeys.name: "",
            ArchivingKeys.URI: "",
            ArchivingKeys.twitterName: "",
            ArchivingKeys.domain: "",
            ArchivingKeys.pages: [
                Page().dictionary
            ]
        ])
    }
    
    // Archiving
    public var manifest: Array<String> {
        get {
            var manifest = [String]()
            for page in self.pages {
                manifest.extend(page.manifest)
            }
            if (!self.URI.isEmpty) {
                manifest.append(self.URI)
            }
            return manifest
        }
    }
    
    public var dictionary: NSDictionary {
        get {
            var pages: Array<NSDictionary> = []
            for page in self.pages {
                pages.append(page.dictionary)
            }
            return [
                ArchivingKeys.name: self.name,
                ArchivingKeys.URI: self.URI,
                ArchivingKeys.twitterName: self.twitterName,
                ArchivingKeys.domain: self.domain,
                ArchivingKeys.pages: pages
            ]
        }
    }
    
    public required init(dictionary: NSDictionary) {
        self.name = dictionary[ArchivingKeys.name] as String
        self.URI = dictionary[ArchivingKeys.URI] as String
        self.twitterName = dictionary[ArchivingKeys.twitterName] as String
        self.domain = dictionary[ArchivingKeys.domain] as String
        self.pages = []
        for page in dictionary[ArchivingKeys.pages] as Array<NSDictionary> {
            self.pages.append(Page(dictionary: page as NSDictionary))
        }
    }
}